hail/src/main/scala/is/hail/backend/service/ServiceBackend.scala

```diff
@@ -152,130 +156,61 @@ class ServiceBackend(
-            JString(root),
-            JString(s"$i"),
-            JString(s"$n"),
-          )),
-          "type" -> JString("jvm"),
-          "profile" -> JBool(backendContext.profile),
-        ),
-        "attributes" -> JObject(
-          "name" -> JString(s"${name}_stage${stageCount}_${stageIdentifier}_job$i")
-        ),
-        "resources" -> resources,
-        "regions" -> JArray(backendContext.regions.map(JString).toList),
-        "cloudfuse" -> JArray(backendContext.cloudfuseConfig.map { config =>
-          JObject(
-            "bucket" -> JString(config.bucket),
-            "mount_path" -> JString(config.mount_path),
-            "read_only" -> JBool(config.read_only),
+    val defaultJob =
+      JobRequest(
+        always_run = false,
+        process = null,
+        resources = Some(
+          JobResources(
+            preemptible = true,
+            cpu = Some(backendContext.workerCores).filter(_ != "None"),
+            memory = Some(backendContext.workerMemory).filter(_ != "None"),
+            storage = Some(backendContext.storageRequirement).filter(_ != "0Gi"),
           )
-        }.toList),
+        ),
+        regions = Some(backendContext.regions).filter(_.nonEmpty),
+        cloudfuse = Some(backendContext.cloudfuseConfig).filter(_.nonEmpty),
       )
-    }
 
-    log.info(s"parallelizeAndComputeWithIndex: $token: running job")
-
-    val (batchId, (updateId, jobGroupId)) = curBatchId match {
-      case Some(id) =>
-        (id, batchClient.update(id, token, workerJobGroup, jobs))
-      case None =>
-        val batchId = batchClient.create(
-          JObject(
-            "billing_project" -> JString(backendContext.billingProject),
-            "n_jobs" -> JInt(n),
-            "token" -> JString(token),
-            "attributes" -> JObject("name" -> JString(name + "_" + stageCount)),
+    val jobs =
+      collection.indices.map { i =>
+        defaultJob.copy(
+          attributes = Map("name" -> s"${name}_stage${stageCount}_${stageIdentifier}_job$i"),
+          process = defaultProcess.copy(
+            command = Array(Main.WORKER, root, s"$i", s"${collection.length}")
           ),
-          jobs,
         )
-        (batchId, (1L, 1L))
-    }
+      }
 
-    val batch = batchClient.waitForJobGroup(batchId, jobGroupId)
+    val jobGroupId =
+      batchClient.newJobGroup(
+        JobGroupRequest(
+          batch_id = batchConfig.batchId,
+          absolute_parent_id = batchConfig.jobGroupId,
+          token = token,
+          attributes = Map("name" -> stageIdentifier),
+          jobs = jobs,
+        )
+      )
 
     stageCount += 1
-    implicit val formats: Formats = DefaultFormats
-    val batchState = (batch \ "state").extract[String]
-    if (batchState == "failed") {
-      throw new HailBatchFailure(s"Update $updateId for batch $batchId failed")
-    }
 
-    (token, root, n)
+    Thread.sleep(600) // it is not possible for the batch to be finished in less than 600ms
```

Comment:
Because the sleep in wait for job group assumes you've just created the job group and I prefer to encode that knowledge in the control flow for creating and polling job groups.